dask/bytes/core.py

```diff
@@ -374,47 +318,93 @@ def open_files(urlpath, compression=None, mode='rb', encoding='utf8',
 
-def get_fs_paths_myopen(urlpath, compression, mode, encoding='utf8',
-                        errors='strict', num=1, name_function=None, **kwargs):
+
+def infer_options(urlpath):
     if hasattr(urlpath, 'name'):
         # deal with pathlib.Path objects - must be local
         urlpath = str(urlpath)
         ispath = True
     else:
         ispath = False
-    if isinstance(urlpath, (str, unicode)):
-        myopen = OpenFileCreator(urlpath, compression, text='b' not in mode,
-                                 encoding=encoding, errors=errors, **kwargs)
-        if ispath and myopen.protocol != 'file':
-            raise ValueError("Only use pathlib.Path with local files.")
+
+    options = infer_storage_options(urlpath)
+    protocol = options.pop('protocol')
+    urlpath = options.pop('path')
+
+    if ispath and protocol != 'file':
+        raise ValueError("Only use pathlib.Path with local files.")
+
+    return urlpath, protocol, options
+
+
+def get_fs_token_paths(urlpath, mode='rb', num=1, name_function=None,
+                       storage_options=None):
+    """Filesystem, deterministic token, and paths from a urlpath and options.
+
+    Parameters
+    ----------
+    urlpath : string
+        Absolute or relative filepath, URL (may include protocols like
+        ``s3://``), or globstring pointing to data.
+    mode : str, optional
+        Mode in which to open files.
+    num : int, optional
+        If opening in writing mode, number of files we expect to create.
+    name_function : callable, optional
+        If opening in writing mode, this callable is used to generate path
+        names. Names are generated for each partition by
+        ``urlpath.replace('*', name_function(partition_index))``.
+    storage_options : dict, optional
+        Additional keywords to pass to the filesystem class.
+    """
+    if isinstance(urlpath, (list, tuple)):
+        if not urlpath:
+            raise ValueError("empty urlpath")
+        paths, protocols, options_itbl = zip(*map(infer_options, urlpath))
+        protocol = protocols[0]
+        options = options_itbl[0]
+        if not (all(p == protocol for p in protocols) and
+                all(o == options for o in options_itbl)):
+            raise ValueError("When specifying a list of paths, all paths must "
+                             "share the same protocol and options")
+        update_storage_options(options, storage_options)
+        paths = list(paths)
+
+        fs, fs_token = get_fs(protocol, options)
+
+    elif isinstance(urlpath, (str, unicode)) or hasattr(urlpath, 'name'):
+        urlpath, protocol, options = infer_options(urlpath)
+        update_storage_options(options, storage_options)
+
+        fs, fs_token = get_fs(protocol, options)
+
         if 'w' in mode:
             paths = _expand_paths(urlpath, name_function, num)
         elif "*" in urlpath:
-            paths = myopen.fs.glob(urlpath)
+            paths = sorted(fs.glob(urlpath))
```

Comment:
Is sorting always a good idea?

Reply:
I see that we're accustomed to sorting below.  So nevermind on this.

Reply:
Ideally `glob` should be sorted, but I don't think we can depend on all backends to do that. I would say that we definitely require that the order is deterministic.
I wonder, should other glob wildcards be checked for (?, [..]) in this blanch?

Reply:
Yes, we should really be sorting everywhere. The `glob` in local filesystem calls `sorted`, which should maybe be removed (idk). In general returning sorted input is not a requirement of `glob` implementations, but we do require a deterministic order.

> I wonder, should other glob wildcards be checked for (?, [..]) in this blanch?

I'd prefer to handle this in a later PR.